sql/lease.go

```diff
@@ -0,0 +1,540 @@
+	if log.V(3) {
+		log.Infof("release: %s", prettyKey(leaseKey, 0))
+	}
+	return s.db.Txn(func(txn *client.Txn) error {
+		b := &client.Batch{}
+		b.Del(leaseKey)
+		txn.SetSystemDBTrigger()
+		return txn.CommitInBatch(b)
+	})
+}
+
+// Publish a new version of a table descriptor. The update closure may be
+// called multiple times if retries occur: make sure it does not have side
+// effects.
+func (s LeaseStore) Publish(tableID ID, update func(*TableDescriptor) error) error {
+	desc := &TableDescriptor{}
+	descKey := MakeDescMetadataKey(tableID)
+
+	retryOpts := retry.Options{
+		InitialBackoff: 20 * time.Millisecond,
+		MaxBackoff:     2 * time.Second,
+		Multiplier:     2,
+	}
+
+	for r := retry.Start(retryOpts); r.Next(); {
+		// Get the current version of the table descriptor non-transactionally.
+		//
+		// TODO(pmattis): Do an inconsistent read here?
+		if err := s.db.GetProto(descKey, desc); err != nil {
+			return err
+		}
+		// Check to see if there are any leases that still exist on the previous
+		// version of the descriptor.
+		now := s.clock.Now()
+		count, err := s.countLeases(desc.ID, desc.Version-1, now.WallTime)
+		if err != nil {
+			return err
+		}
+		if count != 0 {
+			log.Infof("publish (count leases): descID=%d version=%d count=%d", desc.ID, desc.Version-1, count)
+			continue
+		}
+
+		// At this point, desc.Version is the only version of the descriptor that
+		// has leases outstanding. Lease acquisition (see acquire()) maintains the
+		// invariant that no new leases for desc.Version-1 will be granted once
+		// desc.Version exists.
+		expectedVersion := desc.Version
+		err = s.db.Txn(func(txn *client.Txn) error {
+			// Re-read the current version of the table descriptor, this time
+			// transactionally.
+			if err := txn.GetProto(descKey, desc); err != nil {
+				return err
+			}
+			if expectedVersion != desc.Version {
+				// The version changed out from under us. Someone else must be
+				// performing a schema change operation.
+				if log.V(3) {
+					log.Infof("publish (version changed): %d != %d", expectedVersion, desc.Version)
+				}
+				return errLeaseVersionChanged
+			}
+
+			// Update the descriptor.
+			if err := update(desc); err != nil {
+				return err
+			}
+
+			// Bump the version and modification time.
+			desc.Version = desc.Version + 1
+			desc.ModificationTime = now
+			if log.V(3) {
+				log.Infof("publish: descID=%d version=%d mtime=%s", desc.ID, desc.Version, now.GoTime())
+			}
+
+			// Write the updated descriptor.
+			b := &client.Batch{}
+			b.Put(descKey, desc)
+			txn.SetSystemDBTrigger()
+			return txn.CommitInBatch(b)
+		})
+
+		if err != errLeaseVersionChanged {
+			return err
+		}
+	}
+
+	panic("not reached")
+}
+
+// countLeases returns the number of unexpired leases for a particular version
+// of a descriptor.
+func (s LeaseStore) countLeases(descID ID, version uint32, expiration int64) (int, error) {
```

Comment:
True. This isn't used during testing, but I do log the value. With `hasActiveLease` I could change the scan below to limit to a single key.